<a href="https://colab.research.google.com/github/henriquepaes1/churn/blob/main/Projeto_Churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---
<img alt="Colaboratory logo" width="15%" src="https://raw.githubusercontent.com/carlosfab/escola-data-science/master/img/novo_logo_bg_claro.png">

#### **Data Science na Prática 3.0**
*by [sigmoidal.ai](https://sigmoidal.ai)*  

---

# Churn Rate e Machine Learning

*Churn rate*, ou simplesmente *churn*, representa a taxa de evasão da sua base de clientes. Em serviços como Spotify ou Netflix, ela representaria a taxa de cancelamento de assinaturas. A análise dessa métrica é de extrema importância para as equipes de venda, pois dão respostas interessantes à respeito dos pontos fracos de seus serviços e qual é o perfil dos clientes mais fiéis.

<p align=center>
<img src="https://raw.githubusercontent.com/carlosfab/dsnp2/master/img/churnrate.jpg" width="60%"></p>


*Churn* também pode ser usado para identificar potenciais cancelamentos, com um tempo de antecedência, e promover ações direcionadas para tentar reter tais clientes, em geral, quando um cliente aparenta estar próximo de romper vínculo com a empresa, cabe ao time de vendas uma abordagem mais próxima para entender e resolver as dores desse cliente. Essa métrica deve receber atenção pelo fato de que o Custo de Aquisição de Cliente (CAC) é normalmente mais alto que o custo para mantê-los. Ou seja, um alto valor para o *churn rate* é o que não desejamos. É nesse momento que os times de venda das empresas mundo afora recorrem às técnicas de Machine Learning para entender melhor sua base de clientes e como seus serviços a atinge, e é isso que esse artigo traz, um modelo de Machine Learning capaz de dizer se um cliente irá ou não cancelar um serviço.

## Aquisição dos Dados

Os dados utilizados neste projeto foram originalmente disponibilizados na [plataforma de ensino da IBM Developer](https://developer.ibm.com/technologies/data-science/patterns/predict-customer-churn-using-watson-studio-and-jupyter-notebooks/#), e tratam de um problema típico de uma companhia de telecomunicações. O *dataset* completo pode ser encontrado [neste link](https://raw.githubusercontent.com/carlosfab/dsnp2/master/datasets/WA_Fn-UseC_-Telco-Customer-Churn.csv).

Apesar de não haver informações explícitas disponíves, os nomes das colunas permitem um entendimento a respeito do problema.

In [1]:
# importando bibliotecas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


# importar os dados
DATA_PATH = "https://raw.githubusercontent.com/carlosfab/dsnp2/master/datasets/WA_Fn-UseC_-Telco-Customer-Churn.csv"
df = pd.read_csv(DATA_PATH)
# ver as 5 primeiras entradas
df.head()

customerID  gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
0  7590-VHVEG  Female              0  ...          29.85        29.85     No
1  5575-GNVDE    Male              0  ...          56.95       1889.5     No
2  3668-QPYBK    Male              0  ...          53.85       108.15    Yes
3  7795-CFOCW    Male              0  ...          42.30      1840.75     No
4  9237-HQITU  Female              0  ...          70.70       151.65    Yes

[5 rows x 21 columns]

In [2]:
df.shape

(7043, 21)

## Explorando o dataset 
Antes de começarmos as análises, é importante entender como nosso dataset está disposto, seus tipos e afins.

### Tipos das variáveis

In [3]:
# Verificar o tipo de dado salvo em cada coluna
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [4]:
# Lidar com valores ausentes
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Algumas considerações importantes sobre as variáveis das colunas:

* A coluna tenure indica há quantos meses o cliente assina o serviço
* A coluna TotalCharges tem entrada igual a um espaço em branco quando o cliente ainda não completou seu primeiro mês de uso. Portanto, é necessário mudar essas entradas para 0.

## Entendendo o problema
Dada a introdução prévia a respeito do problema, vamos começar os estudos para entender como extrair as melhores informações do conjunto de dados. Nesse estudo, vamos considerar que quanto mais serviços um cliente contrata, maior a "fidelidade", consequentemente, menos chance desse cliente deixar de contratar os serviços da empresa. Assim, é possível definir uma espécie de "hierarquia", quando levando em consideração os serviços assinados por um dado cliente.

## Removendo colunas irrelevantes para a análise
Existem algumas colunas que não dizem respeito ao uso dos serviços fornecidos pela empresa, que não são relevantes para a previsão de evasão de cliente proposta nesse trabalho.

In [5]:
df.drop(['customerID'], axis=1, inplace=True)

Para análise mais imediata, desconsiderando os descontos oferecidos por tempo de assinatura (que podem ser notados ao multiplicar o número de meses de assinatura pelo valor mensal pago), é válido excluir a coluna "TotalCharges", que indicam quanto o cliente gastou em todo o tempo de contrato, a fim de entender como o valor da parcela influencia no cancelamento do cliente.

In [6]:
df.drop('TotalCharges', axis=1, inplace=True)


## Lidando com as variáveis categóricas
No nosso dataset, foi possível observar muitas variáveis do tipo object, que prejudicam o modelo matematicamente. Por isso, é necessário fazer algumas alterações nos dados.

Inicialmente, modificaremos as colunas *churn* (1 se a empresa perdeu o cliente, 0 se não) e *tenure* (quantos meses o cliente possui vínculo com a empresa), transformando-as para um tipo numérico. 

In [7]:
df.Churn.replace({"Yes": '1', "No": '0'}, inplace=True)
df.Churn = pd.to_numeric(df.Churn) # houve churn

# Alterar o tipo das colunas para numérico quando aplicável
df.tenure = pd.to_numeric(df.tenure) # meses de assinatura

Nesse momento, nosso dataset ainda conta com muitas variáveis categóricas. Para lidarmos com elas corretamente, sem comprometer nossa análise matematicamente (adicionar "pesos" desnecessários a algumas variáveis), vamos utilizar a técnica de OneHotEncoding, que separará as respostas possíveis para os campos em colunas, com valores 1 ou 0, indicando se o cliente contrata ou não determinado serviço.


In [8]:
df = pd.get_dummies(df)


In [9]:
df.dtypes

SeniorCitizen                                int64
tenure                                       int64
MonthlyCharges                             float64
Churn                                        int64
gender_Female                                uint8
gender_Male                                  uint8
Partner_No                                   uint8
Partner_Yes                                  uint8
Dependents_No                                uint8
Dependents_Yes                               uint8
PhoneService_No                              uint8
PhoneService_Yes                             uint8
MultipleLines_No                             uint8
MultipleLines_No phone service               uint8
MultipleLines_Yes                            uint8
InternetService_DSL                          uint8
InternetService_Fiber optic                  uint8
InternetService_No                           uint8
OnlineSecurity_No                            uint8
OnlineSecurity_No internet serv

In [10]:
df.head()

SeniorCitizen  ...  PaymentMethod_Mailed check
0              0  ...                           0
1              0  ...                           1
2              0  ...                           1
3              0  ...                           0
4              0  ...                           0

[5 rows x 45 columns]

Agora que todas as variáveis são numéricas, é possível dar o próximo passo na análise.

## Normalizando as variáveis
Mesmo com todas as variáveis categóricas sendo devidamente trabalhadas, algumas colunas ainda podem enviezar nosso estudo por diversos motivos. No nosso caso, as colunas "tenure" e "MonthlyCharge" possuem valores muito maiores que os zeros e uns do resto do dataset, isso por que possuem uma escala própria (meses e dólares, respectivamente), isso exige uma *normalização*. Por termos usado o método de OneHotEncoder, queremos que as colunas *tenure* e *MonthlyCharge* também tenham valores entre 0 e 1, para isso, utilizaremos o MinMaxScaler.

In [11]:
X = df.drop('Churn', axis=1)
y = df.Churn

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

X_train, X_test, y_train, y_test = train_test_split(X, y)
scaler = MinMaxScaler()

In [13]:
df_train = X_train.copy()
df_test = X_test.copy()

df_train['MonthlyCharges'] = scaler.fit_transform(np.array(df_train['MonthlyCharges']).reshape(-1, 1))
df_train.tenure = scaler.fit_transform(np.array(df_train.tenure).reshape(-1, 1))
df_test['MonthlyCharges'] = scaler.fit_transform(np.array(df_test['MonthlyCharges']).reshape(-1, 1))
df_test.tenure = scaler.fit_transform(np.array(df_test.tenure).reshape(-1, 1))

## Balanceamento dos dados

In [14]:
df.Churn.value_counts()

0    5174
1    1869
Name: Churn, dtype: int64

Aqui, pode-se perceber que trata-se de um dataset desbalanceado e que, algumas entradas devem ser removidas para que o modelo funcione melhor. Vamos balancear os dados removendo entradas da classe majoritária.

In [15]:
from imblearn.under_sampling import RandomUnderSampler

In [16]:
rus = RandomUnderSampler()
X_treino, y_treino = rus.fit_resample(df_train, y_train)

In [17]:
y_treino.value_counts()

1    1427
0    1427
Name: Churn, dtype: int64

Agora, com os datasets prontos, podemos começar a criar os modelos.

## Criando os modelos 


Por tratar-se de um problema de classificação, criaremos 2 modelos: um utilizando Regressão Logística e outro utilizando o método dos K-ésimos vizinhos mais próximos.

In [18]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score
import seaborn as sns

In [19]:
# Criando modelo de cross validation para avaliar o recall dos modelos

log = LogisticRegression()
scores_log = cross_val_score(log, X_treino, y_treino, scoring='recall') # Método de Cross-Validation
log.fit(X_treino, y_treino)
pred_log = log.predict(X_test) # usando apenas o dataset de teste (desbalanceado)
print(classification_report(y_test, pred_log))
print(scores_log)
print("Recall médio: {:.2f}%".format(scores_log.mean()*100))


              precision    recall  f1-score   support

           0       0.75      1.00      0.86      1319
           1       0.00      0.00      0.00       442

    accuracy                           0.75      1761
   macro avg       0.37      0.50      0.43      1761
weighted avg       0.56      0.75      0.64      1761

[0.7754386  0.85964912 0.76923077 0.79370629 0.76491228]
Recall médio: 79.26%


/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Pode-se ver que, para a Regressão Logística, os resultados para a variável 1 são baixos, por conta do desbalanceamento do dataset de teste, porém, ainda assim no geral, o modelo obteve bons resultados de precisão e recall. Vamos ver os resultados com outros modelos.

In [20]:
knn = KNeighborsClassifier()
scores_knn = cross_val_score(knn, X_treino, y_treino, scoring='recall') # Método de Cross-Validation
knn.fit(X_treino, y_treino)
pred_knn = knn.predict(X_test) # usando apenas o dataset de teste (desbalanceado)
print(classification_report(y_test, pred_knn))
print(scores_knn)
print("Recall médio: {:.2f}%".format(scores_knn.mean()*100))

              precision    recall  f1-score   support

           0       0.75      0.94      0.83      1319
           1       0.29      0.07      0.12       442

    accuracy                           0.72      1761
   macro avg       0.52      0.51      0.48      1761
weighted avg       0.63      0.72      0.65      1761

[0.81754386 0.83508772 0.81118881 0.7972028  0.79649123]
Recall médio: 81.15%


In [21]:
rf = RandomForestClassifier()
scores_rf = cross_val_score(rf, X_treino, y_treino, scoring='recall') # Método de Cross-Validation
rf.fit(X_treino, y_treino)
pred_rf = rf.predict(X_test) # usando apenas o dataset de teste (desbalanceado)
print(classification_report(y_test, pred_rf))
print(scores_rf)
print("Recall médio: {:.2f}%".format(scores_rf.mean()*100))

              precision    recall  f1-score   support

           0       0.82      0.87      0.84      1319
           1       0.52      0.44      0.48       442

    accuracy                           0.76      1761
   macro avg       0.67      0.65      0.66      1761
weighted avg       0.75      0.76      0.75      1761

[0.72280702 0.76842105 0.77272727 0.75874126 0.72631579]
Recall médio: 74.98%


In [22]:
dt = DecisionTreeClassifier()
scores_dt = cross_val_score(dt, X_treino, y_treino, scoring='recall') # Método de Cross-Validation
dt.fit(X_treino, y_treino)
pred_dt = dt.predict(X_test) # usando apenas o dataset de teste (desbalanceado)
print(classification_report(y_test, pred_dt))
print(scores_dt)
print("Recall médio: {:.2f}%".format(scores_dt.mean()*100))

              precision    recall  f1-score   support

           0       0.74      0.84      0.79      1319
           1       0.19      0.11      0.14       442

    accuracy                           0.66      1761
   macro avg       0.46      0.47      0.46      1761
weighted avg       0.60      0.66      0.62      1761

[0.67017544 0.72982456 0.67482517 0.66783217 0.6877193 ]
Recall médio: 68.61%


Com esses dados, podemos concluir que o modelo que melhor performou foi aquele que utilizou o algoritmo dos K-ésimos vizinhos mais próximos, obtendo recall médio de 81%. Além disso, podemos concluir que o balanceamento do dataset e de suma importância para determinação dos resultados, como pudemos notar, ao analisar o recall somente para quando o churn ocorre, sem o metódo de Cross-validation, temos a impressão de que o modelo é pior do que ele realmente é.